A71-xds-fallback.md

```diff
@@ -21,81 +21,79 @@ obtained.
 ## Background
 
 xDS (also known as xDS Discovery Service) is a suite of APIs for discovering
-and subscribing to the configuration of a server mesh. Orignially created by
-for the [Envoy Proxy](http://envoyproxy.io) project, the xDS protocol is
-evolving into an industry standard. gRPC introduced a number of features that
-can be configured using this protocol.
-
-Components of the server mesh communicate with xDS server(s) (also known as
-a "control plane") to obtain the initial configuration and to receive updates
-to the configuration.
-
-An xDS configuration consists of a number of "resources" that can be queried
-separately and may change independently from one another.
-
-Even a brief downtime of the xDS control plane may cause significant disruption
-in the service mesh inter-component connectivity and result in wider outages.
-Examples include:
-
-* Services becoming unavailable as they fail to access downstream dependencies.
-* Traffic being routed incorrectly failing to properly balance the load between
-the endpoints.
-* Security risks may be introduced as outdated credentials are used for
-communications between the servers.
+and subscribing to the configuration of a server mesh. Even a brief downtime
+of the xDS control plane may cause significant disruption in the service mesh
+inter-component connectivity and result in wider outages.
 
 Current xDS implementation in gRPC uses a single global instance of
 the XdsClient class to communicate with the xDS control plane. This instance
 is shared across all channels to reduce overhead by providing a shared cache
-for xDS resources as well as reducing a number of connections to xDS servers.
+for xDS resources as well as reducing the number of connections to xDS servers.
 
 ### Related Proposals: 
-* [A27: xDS-Based Global Load Balancing](A27)
-* [A36: xDS-Enabled Servers](A36)
-* [A57: XdsClient Failure Mode Behavior](A57)
+* [A27: xDS-Based Global Load Balancing][A27]
+* [A36: xDS-Enabled Servers][A36]
+* [A47: xDS Federation][A47]
+* [A57: XdsClient Failure Mode Behavior][A57]
 
 [A27]: A27-xds-global-load-balancing.md
 [A36]: A36-xds-for-servers.md
+[A47]: A47-xds-federation.md
 [A57]: A57-xds-client-failure-mode-behavior.md
 
 ## Proposal
 
-gRPC code should use fallback iff both:
+gRPC code should use fallback if the following conditions hold:
 
-* primary server not reachable
-* at least one watcher exists for a resource that is not cached (where
-   "does not exist" counts as cached).
+* There had been a failure during resource retrieval, as described in [A57]:
+    - Connection to the data plane fails. 
+    - The ADS stream being closed before the first server response had been
+      received.
+* At least one watcher exists for a resource that is not cached. Resources are
+    considered cached if:
+    - A resource had been successfully retried from the xDS server.
```

Comment:
These three bullets are very confusing and I think inaccurate in a few ways.  I suggest writing this as follows:

A resource is considered not to be cached if the server has not yet delivered it to the client.  Note that a resource *is* considered to be cached in all of the following cases:
- The resource has been returned by the server, regardless of whether the resource was valid or invalid (i.e., regardless of whether we ACKed or NACKed it).
- The resource does not exist, either because the server did not return it within the [timeout period](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#knowing-when-a-requested-resource-does-not-exist), or because [the server sent an LDS or CDS response that no longer includes the resource](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#deleting-resources).

Reply:
I believe we should have a caveat on the "server did not return it within the timeout period" to require that something was returned since the timeout.  Otherwise, the server becoming unavailable could be falsely interpreted as the resource not existing.

Reply:
I think the requirements of [gRFC A57](https://github.com/grpc/proposal/blob/master/A57-xds-client-failure-mode-behavior.md) already clearly differentiate between those cases.  The does-not-exist timer should not be running when we're not actually in contact with the server.